diff -rupN a/certs/extract-cert.c b/certs/extract-cert.c
--- a/certs/extract-cert.c	2026-04-15 11:45:08.000000000 +0200
+++ b/certs/extract-cert.c	2026-04-15 12:27:23.770890362 +0200
@@ -41,7 +41,9 @@ void format(void)
 	exit(2);
 }
 
+#ifdef USE_PKCS11_ENGINE
 static const char *key_pass;
+#endif
 static BIO *wb;
 static char *cert_dst;
 static bool verbose;
@@ -133,7 +135,9 @@ int main(int argc, char **argv)
 	if (verbose_env && strchr(verbose_env, '1'))
 		verbose = true;
 
-        key_pass = getenv("KBUILD_SIGN_PIN");
+#ifdef USE_PKCS11_ENGINE
+	key_pass = getenv("KBUILD_SIGN_PIN");
+#endif
 
 	if (argc != 3)
 		format();
