diff -rupN a/io_uring/fdinfo.c b/io_uring/fdinfo.c
--- a/io_uring/fdinfo.c	2025-10-26 08:07:10.000000000 +0100
+++ b/io_uring/fdinfo.c	2025-10-26 09:38:13.437069573 +0100
@@ -91,7 +91,7 @@ static void __io_uring_show_fdinfo(struc
 	sq_entries = min(sq_tail - sq_head, ctx->sq_entries);
 	for (i = 0; i < sq_entries; i++) {
 		unsigned int entry = i + sq_head;
-		struct io_uring_sqe *sqe;
+		struct io_uring_sqe *sqe = NULL;
 		unsigned int sq_idx;
 
 		if (ctx->flags & IORING_SETUP_NO_SQARRAY)
