diff -rupN a/drivers/hid/hid-logitech-hidpp.c b/drivers/hid/hid-logitech-hidpp.c
--- a/drivers/hid/hid-logitech-hidpp.c	2025-12-13 13:26:11.357317157 +0100
+++ b/drivers/hid/hid-logitech-hidpp.c	2025-12-13 13:36:51.880393435 +0100
@@ -980,7 +980,7 @@ static int hidpp_root_get_protocol_versi
 		return -EIO;
 
 	if (ret > 0) {
-		hid_err(hidpp->hid_dev, "%s: received protocol error 0x%02x\n",
+		hid_dbg(hidpp->hid_dev, "%s: received protocol error 0x%02x\n",
 			__func__, ret);
 		return -EPROTO;
 	}
