diff -rupN a/kernel/workqueue.c b/kernel/workqueue.c
--- a/kernel/workqueue.c	2025-07-25 12:35:28.000000000 +0200
+++ b/kernel/workqueue.c	2025-07-27 22:57:33.866920742 +0200
@@ -1336,7 +1336,7 @@ restart:
 		if (wq_cpu_intensive_warning_thresh &&
 		    cnt >= wq_cpu_intensive_warning_thresh &&
 		    is_power_of_2(cnt + 1 - wq_cpu_intensive_warning_thresh))
-			printk_deferred(KERN_WARNING "workqueue: %ps hogged CPU for >%luus %llu times, consider switching to WQ_UNBOUND\n",
+			printk_deferred(KERN_DEBUG "workqueue: %ps hogged CPU for >%luus %llu times, consider switching to WQ_UNBOUND\n",
 					ent->func, wq_cpu_intensive_thresh_us,
 					atomic64_read(&ent->cnt));
 		return;
