diff -rupN a/drivers/gpu/drm/drm_file.c b/drivers/gpu/drm/drm_file.c
--- a/drivers/gpu/drm/drm_file.c	2024-12-28 19:34:58.000000000 +0100
+++ b/drivers/gpu/drm/drm_file.c	2024-12-28 20:13:54.775139385 +0100
@@ -314,8 +314,6 @@ int drm_open_helper(struct file *filp, s
 	if (dev->switch_power_state != DRM_SWITCH_POWER_ON &&
 	    dev->switch_power_state != DRM_SWITCH_POWER_DYNAMIC_OFF)
 		return -EINVAL;
-	if (WARN_ON_ONCE(!(filp->f_op->fop_flags & FOP_UNSIGNED_OFFSET)))
-		return -EINVAL;
 
 	drm_dbg_core(dev, "comm=\"%s\", pid=%d, minor=%d\n",
 		     current->comm, task_pid_nr(current), minor->index);
@@ -333,6 +331,7 @@ int drm_open_helper(struct file *filp, s
 	}
 
 	filp->private_data = priv;
+	filp->f_mode |= FOP_UNSIGNED_OFFSET;
 	priv->filp = filp;
 
 	mutex_lock(&dev->filelist_mutex);
