diff -rupN a/init/Kconfig b/init/Kconfig
--- a/init/Kconfig	2025-01-12 13:13:12.427398709 +0100
+++ b/init/Kconfig	2025-01-12 13:14:05.139918082 +0100
@@ -1482,6 +1482,12 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 	  with the "-O2" compiler flag for best performance and most
 	  helpful compile-time warnings.
 
+config CC_OPTIMIZE_FOR_PERFORMANCE_O3
+	bool "Optimize more for performance (-O3)"
+	help
+	  Choosing this option will pass "-O3" to your compiler to optimize
+	  the kernel yet more for performance.
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size (-Os)"
 	help
diff -rupN a/Makefile b/Makefile
--- a/Makefile	2025-01-12 13:13:12.427398709 +0100
+++ b/Makefile	2025-01-12 13:14:05.139918082 +0100
@@ -860,6 +860,9 @@ KBUILD_CFLAGS	+= -fno-delete-null-pointe
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS += -O2
 KBUILD_RUSTFLAGS += -Copt-level=2
+else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
+KBUILD_CFLAGS += -O3
+KBUILD_RUSTFLAGS += -Copt-level=3
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
 KBUILD_RUSTFLAGS += -Copt-level=s
