diff -rupN a/arch/x86/Makefile b/arch/x86/Makefile
--- a/arch/x86/Makefile	2025-03-22 09:51:12.000000000 +0100
+++ b/arch/x86/Makefile	2025-03-22 12:30:57.494928676 +0100
@@ -172,8 +172,8 @@ else
 	KBUILD_CFLAGS += $(call cc-option,-mskip-rax-setup)
 
 ifdef CONFIG_X86_NATIVE_CPU
-        KBUILD_CFLAGS += -march=native
-        KBUILD_RUSTFLAGS += -Ctarget-cpu=native
+        KBUILD_CFLAGS += -march=skylake
+        KBUILD_RUSTFLAGS += -Ctarget-cpu=skylake
 else
         KBUILD_CFLAGS += -march=x86-64 -mtune=generic
         KBUILD_RUSTFLAGS += -Ctarget-cpu=x86-64 -Ztune-cpu=generic
