diff -rupN a/drivers/hid/hid-logitech-hidpp.c b/drivers/hid/hid-logitech-hidpp.c
--- a/drivers/hid/hid-logitech-hidpp.c	2024-12-08 23:03:39.000000000 +0100
+++ b/drivers/hid/hid-logitech-hidpp.c	2024-12-09 15:23:17.492844383 +0100
@@ -973,7 +973,7 @@ static int hidpp_root_get_protocol_versi
 		return -EIO;
 
 	if (ret > 0) {
-		hid_err(hidpp->hid_dev, "%s: received protocol error 0x%02x\n",
+		hid_dbg(hidpp->hid_dev, "%s: received protocol error 0x%02x\n",
 			__func__, ret);
 		return -EPROTO;
 	}
