diff -rupN a/nvidia-drm/nvidia-drm-helper.h b/nvidia-drm/nvidia-drm-helper.h
--- a/nvidia-drm/nvidia-drm-helper.h	2025-09-23 11:40:05.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-helper.h	2025-10-22 10:40:28.965256185 +0200
@@ -196,7 +196,7 @@ int nv_drm_atomic_helper_disable_all(str
        for ((__i) = 0;                                                      \
             (__i) < (__state)->num_connector &&                             \
             ((connector) = (__state)->connectors[__i].ptr,                  \
-            (connector_state) = (__state)->connectors[__i].state, 1);       \
+            (connector_state) = (__state)->connectors[__i].new_state, 1);       \
             (__i)++)                                                        \
                for_each_if (connector)
 #else
@@ -223,7 +223,7 @@ int nv_drm_atomic_helper_disable_all(str
        for ((__i) = 0;                                                \
             (__i) < (__state)->dev->mode_config.num_crtc &&           \
             ((crtc) = (__state)->crtcs[__i].ptr,                      \
-            (crtc_state) = (__state)->crtcs[__i].state, 1);           \
+            (crtc_state) = (__state)->crtcs[__i].new_state, 1);           \
             (__i)++)                                                  \
                for_each_if (crtc_state)
 #else
@@ -248,7 +248,7 @@ int nv_drm_atomic_helper_disable_all(str
        for ((__i) = 0;                                                   \
             (__i) < (__state)->dev->mode_config.num_total_plane &&       \
             ((plane) = (__state)->planes[__i].ptr,                       \
-            (plane_state) = (__state)->planes[__i].state, 1);            \
+            (plane_state) = (__state)->planes[__i].new_state, 1);            \
             (__i)++)                                                     \
                for_each_if (plane_state)
 #else
