diff -rupN a/Makefile b/Makefile
--- a/Makefile	2026-03-31 15:18:33.929571291 +0200
+++ b/Makefile	2026-03-31 15:18:41.686862264 +0200
@@ -907,7 +907,7 @@ endif # need-config
 KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
-KBUILD_CFLAGS += -O2
+KBUILD_CFLAGS += -O3
 KBUILD_RUSTFLAGS += -Copt-level=2
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
