diff -rupN a/Makefile b/Makefile
--- a/Makefile	2026-04-15 14:47:27.441194997 +0200
+++ b/Makefile	2026-04-15 14:47:37.549893328 +0200
@@ -923,7 +923,7 @@ endif # need-config
 KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
-KBUILD_CFLAGS += -O2
+KBUILD_CFLAGS += -O3
 KBUILD_RUSTFLAGS += -Copt-level=2
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
