diff -rupN a/drivers/gpu/drm/Kconfig b/drivers/gpu/drm/Kconfig
--- a/drivers/gpu/drm/Kconfig	2026-02-25 17:51:28.000000000 +0100
+++ b/drivers/gpu/drm/Kconfig	2026-02-25 19:04:34.098778757 +0100
@@ -186,7 +186,7 @@ config DRM_LOAD_EDID_FIRMWARE
 source "drivers/gpu/drm/display/Kconfig"
 
 config DRM_TTM
-	tristate
+	tristate "TTM"
 	depends on DRM && MMU
 	select SHMEM
 	help
@@ -231,7 +231,7 @@ config DRM_VRAM_HELPER
 	  Helpers for VRAM memory management
 
 config DRM_TTM_HELPER
-	tristate
+	tristate "TTM_HELPER"
 	depends on DRM
 	select DRM_TTM
 	select DRM_KMS_HELPER if DRM_FBDEV_EMULATION
