diff -rupN a/arch/x86/Makefile b/arch/x86/Makefile
--- a/arch/x86/Makefile	2026-03-06 16:41:24.000000000 +0100
+++ b/arch/x86/Makefile	2026-03-08 21:40:41.781820491 +0100
@@ -161,7 +161,7 @@ else
 	KBUILD_CFLAGS += -mskip-rax-setup
 
 ifdef CONFIG_X86_NATIVE_CPU
-        KBUILD_CFLAGS += -march=native
+        KBUILD_CFLAGS += -march=skylake
         KBUILD_RUSTFLAGS += -Ctarget-cpu=native
 else
         KBUILD_CFLAGS += -march=x86-64 -mtune=generic
