diff -rupN a/Makefile b/Makefile
--- a/Makefile	2026-03-16 20:11:16.807399791 +0100
+++ b/Makefile	2026-03-16 20:21:55.198080990 +0100
@@ -902,7 +902,7 @@ endif # need-config
 KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
-KBUILD_CFLAGS += -O2
+KBUILD_CFLAGS += -O3
 KBUILD_RUSTFLAGS += -Copt-level=2
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
