diff -rupN a/arch/x86/Makefile b/arch/x86/Makefile
--- a/arch/x86/Makefile	2025-12-29 10:20:05.000000000 +0100
+++ b/arch/x86/Makefile	2025-12-29 11:02:10.109358679 +0100
@@ -165,7 +165,7 @@ else
 	KBUILD_CFLAGS += -mskip-rax-setup
 
 ifdef CONFIG_X86_NATIVE_CPU
-        KBUILD_CFLAGS += -march=native
+        KBUILD_CFLAGS += -march=skylake
         KBUILD_RUSTFLAGS += -Ctarget-cpu=native
 else
         KBUILD_CFLAGS += -march=x86-64 -mtune=generic
