diff -rupN a/nvidia-drm/nvidia-drm.c b/nvidia-drm/nvidia-drm.c
--- a/nvidia-drm/nvidia-drm.c	2025-09-23 11:40:05.000000000 +0200
+++ b/nvidia-drm/nvidia-drm.c	2025-11-05 11:12:53.667282587 +0100
@@ -26,6 +26,7 @@
 
 #include "nvidia-drm-priv.h"
 #include "nvidia-drm-drv.h"
+#include <drm/drm_print.h>
 
 static struct NvKmsKapiFunctionsTable nvKmsFuncsTable = {
     .versionString = NV_VERSION_STRING,
diff -rupN a/nvidia-drm/nvidia-drm-connector.c b/nvidia-drm/nvidia-drm-connector.c
--- a/nvidia-drm/nvidia-drm-connector.c	2025-09-23 11:40:07.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-connector.c	2025-11-05 11:15:47.213364132 +0100
@@ -41,6 +41,7 @@
 #endif
 #include <drm/drm_crtc_helper.h>
 
+#include <drm/drm_print.h>
 #include <drm/drm_atomic.h>
 #include <drm/drm_atomic_helper.h>
 #include <drm/drm_edid.h>
diff -rupN a/nvidia-drm/nvidia-drm-crtc.c b/nvidia-drm/nvidia-drm-crtc.c
--- a/nvidia-drm/nvidia-drm-crtc.c	2025-09-23 11:40:07.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-crtc.c	2025-11-05 11:16:07.589639419 +0100
@@ -40,6 +40,7 @@
 #include <drm/drm_crtc_helper.h>
 #include <drm/drm_plane_helper.h>
 
+#include <drm/drm_print.h>
 #include <drm/drm_atomic.h>
 #include <drm/drm_atomic_helper.h>
 #include <drm/drm_color_mgmt.h>
diff -rupN a/nvidia-drm/nvidia-drm-drv.c b/nvidia-drm/nvidia-drm-drv.c
--- a/nvidia-drm/nvidia-drm-drv.c	2025-09-23 11:40:07.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-drv.c	2025-11-05 11:16:43.053835098 +0100
@@ -50,6 +50,7 @@
 #include <drm/drm_atomic_uapi.h>
 #endif
 
+#include <drm/drm_print.h>
 #include <drm/drm_vblank.h>
 #include <drm/drm_file.h>
 #include <drm/drm_prime.h>
diff -rupN a/nvidia-drm/nvidia-drm-encoder.c b/nvidia-drm/nvidia-drm-encoder.c
--- a/nvidia-drm/nvidia-drm-encoder.c	2025-09-23 11:40:07.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-encoder.c	2025-11-05 11:17:04.325573809 +0100
@@ -43,6 +43,7 @@
 #endif
 #include <drm/drm_crtc_helper.h>
 
+#include <drm/drm_print.h>
 #include <drm/drm_atomic.h>
 #include <drm/drm_atomic_helper.h>
 
diff -rupN a/nvidia-drm/nvidia-drm-fb.c b/nvidia-drm/nvidia-drm-fb.c
--- a/nvidia-drm/nvidia-drm-fb.c	2025-09-23 11:40:08.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-fb.c	2025-11-05 11:17:18.293558246 +0100
@@ -32,6 +32,7 @@
 #include "nvidia-drm-helper.h"
 #include "nvidia-drm-format.h"
 
+#include <drm/drm_print.h>
 #include <drm/drm_crtc_helper.h>
 
 static void __nv_drm_framebuffer_free(struct nv_drm_framebuffer *nv_fb)
diff -rupN a/nvidia-drm/nvidia-drm-fence.c b/nvidia-drm/nvidia-drm-fence.c
--- a/nvidia-drm/nvidia-drm-fence.c	2025-09-23 11:40:08.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-fence.c	2025-11-05 11:17:39.869012442 +0100
@@ -28,6 +28,7 @@
 #include <drm/drmP.h>
 #endif
 
+#include <drm/drm_print.h>
 #include "nvidia-drm-priv.h"
 #include "nvidia-drm-ioctl.h"
 #include "nvidia-drm-gem.h"
diff -rupN a/nvidia-drm/nvidia-drm-gem.h b/nvidia-drm/nvidia-drm-gem.h
--- a/nvidia-drm/nvidia-drm-gem.h	2025-09-23 11:40:06.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-gem.h	2025-11-05 11:18:05.685290997 +0100
@@ -33,6 +33,7 @@
 #include <drm/drmP.h>
 #endif
 
+#include <drm/drm_print.h>
 #include <drm/drm_gem.h>
 
 #include "nvkms-kapi.h"
diff -rupN a/nvidia-drm/nvidia-drm-modeset.c b/nvidia-drm/nvidia-drm-modeset.c
--- a/nvidia-drm/nvidia-drm-modeset.c	2025-09-23 11:40:08.000000000 +0200
+++ b/nvidia-drm/nvidia-drm-modeset.c	2025-11-05 11:18:25.469984980 +0100
@@ -34,6 +34,7 @@
 #include <drm/drmP.h>
 #endif
 
+#include <drm/drm_print.h>
 #include <drm/drm_vblank.h>
 #include <drm/drm_atomic.h>
 #include <drm/drm_atomic_helper.h>
