diff --git a/nvidia/os-interface.c b/nvidia/os-interface.c
index f44bf2d..df5bcb3 100644
--- a/nvidia/os-interface.c
+++ b/nvidia/os-interface.c
@@ -37,6 +37,8 @@
 #include <linux/cpuset.h>
 #include <linux/sys_soc.h>
 
+#include <linux/version.h>
+
 extern char *NVreg_TemporaryFilePath;
 
 #define MAX_ERROR_STRING 528
@@ -2407,7 +2409,15 @@ NvU32 NV_API_CALL os_cgroup_implementation(void)
 
 void* NV_API_CALL os_dmem_cgroup_register_region(NvU64 size, const char *name)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7, 3, 0)
+    struct dmem_cgroup_init init = {
+        .size = size
+    };
+
+    void *region = dmem_cgroup_register_region(&init, name);
+#else
     void *region = dmem_cgroup_register_region(size, name);
+#endif
     if (IS_ERR(region))
     {
         return NULL;
