diff -rupN a/arch/x86/Makefile b/arch/x86/Makefile
--- a/arch/x86/Makefile	2025-07-24 07:23:35.000000000 +0200
+++ b/arch/x86/Makefile	2025-07-24 13:35:19.776176371 +0200
@@ -174,7 +174,7 @@ else
 	KBUILD_CFLAGS += $(call cc-option,-mskip-rax-setup)
 
 ifdef CONFIG_X86_NATIVE_CPU
-        KBUILD_CFLAGS += -march=native
+        KBUILD_CFLAGS += -march=skylake
         KBUILD_RUSTFLAGS += -Ctarget-cpu=native
 else
         KBUILD_CFLAGS += -march=x86-64 -mtune=generic
