diff -rupN a/arch/x86/Makefile b/arch/x86/Makefile
--- a/arch/x86/Makefile	2025-09-05 08:01:32.000000000 +0200
+++ b/arch/x86/Makefile	2025-09-06 13:48:36.697577636 +0200
@@ -163,7 +163,7 @@ else
 	KBUILD_CFLAGS += -mskip-rax-setup
 
 ifdef CONFIG_X86_NATIVE_CPU
-        KBUILD_CFLAGS += -march=native
+        KBUILD_CFLAGS += -march=skylake
         KBUILD_RUSTFLAGS += -Ctarget-cpu=native
 else
         KBUILD_CFLAGS += -march=x86-64 -mtune=generic
