diff --git a/kernel/time/tick-sched.c b/kernel/time/tick-sched.c
index a7893a0..5bb85f5 100644
--- a/kernel/time/tick-sched.c
+++ b/kernel/time/tick-sched.c
@@ -1083,7 +1083,7 @@ static bool report_idle_softirq(void)
 		return false;
 
 	if (ratelimit < 10) {
-		pr_warn("NOHZ tick-stop error: local softirq work is pending, handler #%02x!!!\n",
+		pr_debug("NOHZ tick-stop error: local softirq work is pending, handler #%02x!!!\n",
 			pending);
 		ratelimit++;
 	}
