diff -rupN a/nvidia-drm/nvidia-drm-fb.c b/nvidia-drm/nvidia-drm-fb.c
--- nvidia-580.65.06.ok/nvidia-drm/nvidia-drm-fb.c      2025-07-27 08:41:16.000000000 +0200
+++ nvidia-580.65.06/nvidia-drm/nvidia-drm-fb.c 2025-08-05 12:16:29.512242478 +0200
@@ -242,6 +242,7 @@ fail:
 struct drm_framebuffer *nv_drm_framebuffer_create(
     struct drm_device *dev,
     struct drm_file *file,
+    const struct drm_format_info *info,
     const struct drm_mode_fb_cmd2 *cmd)
 {
     struct nv_drm_device *nv_dev = to_nv_device(dev);
@@ -289,6 +290,7 @@ struct drm_framebuffer *nv_drm_framebuff
     drm_helper_mode_fill_fb_struct(
         dev,
         &nv_fb->base,
+        NULL,
         cmd);
 
     /*
diff -rupN a/nvidia-drm/nvidia-drm-fb.h b/nvidia-drm/nvidia-drm-fb.h
--- nvidia-580.65.06.ok/nvidia-drm/nvidia-drm-fb.h      2025-07-27 08:41:14.000000000 +0200
+++ nvidia-580.65.06/nvidia-drm/nvidia-drm-fb.h 2025-08-05 12:11:48.894145880 +0200
@@ -53,6 +53,7 @@ static inline struct nv_drm_framebuffer
 struct drm_framebuffer *nv_drm_framebuffer_create(
     struct drm_device *dev,
     struct drm_file *file,
+    const struct drm_format_info *info,
     const struct drm_mode_fb_cmd2 *cmd);
 
 #endif /* NV_DRM_AVAILABLE */
