diff -rupN a/nvidia/nv-mmap.c b/nvidia/nv-mmap.c
--- a/nvidia/nv-mmap.c	2025-12-08 14:45:02.000000000 +0100
+++ b/nvidia/nv-mmap.c	2026-01-27 11:35:52.332089753 +0100
@@ -864,13 +864,13 @@ void NV_API_CALL nv_set_safe_to_mmap_loc
 #if !NV_CAN_CALL_VMA_START_WRITE
 static NvBool nv_vma_enter_locked(struct vm_area_struct *vma, NvBool detaching)
 {
-    NvU32 tgt_refcnt = VMA_LOCK_OFFSET;
+    NvU32 tgt_refcnt = VM_REFCNT_EXCLUDE_READERS_FLAG;
     NvBool interrupted = NV_FALSE;
     if (!detaching)
     {
         tgt_refcnt++;
     }
-    if (!refcount_add_not_zero(VMA_LOCK_OFFSET, &vma->vm_refcnt))
+    if (!refcount_add_not_zero(VM_REFCNT_EXCLUDE_READERS_FLAG, &vma->vm_refcnt))
     {
         return NV_FALSE;
     }
@@ -900,7 +900,7 @@ static NvBool nv_vma_enter_locked(struct
     if (interrupted)
     {
         // Clean up on error: release refcount and dep_map
-        refcount_sub_and_test(VMA_LOCK_OFFSET, &vma->vm_refcnt);
+        refcount_sub_and_test(VM_REFCNT_EXCLUDE_READERS_FLAG, &vma->vm_refcnt);
         rwsem_release(&vma->vmlock_dep_map, _RET_IP_);	
         return NV_FALSE;
     }
@@ -916,7 +916,7 @@ void nv_vma_start_write(struct vm_area_s
 {
     NvU32 mm_lock_seq;
     NvBool locked;
-    if (__is_vma_write_locked(vma, &mm_lock_seq))
+    if (__is_vma_write_locked(vma))
         return;
 
     locked = nv_vma_enter_locked(vma, NV_FALSE);
@@ -925,7 +925,7 @@ void nv_vma_start_write(struct vm_area_s
     if (locked)
     {
         NvBool detached;
-        detached = refcount_sub_and_test(VMA_LOCK_OFFSET, &vma->vm_refcnt);
+        detached = refcount_sub_and_test(VM_REFCNT_EXCLUDE_READERS_FLAG, &vma->vm_refcnt);
         rwsem_release(&vma->vmlock_dep_map, _RET_IP_);
         WARN_ON_ONCE(detached);
     }
diff -rupN a/nvidia/os-interface.c b/nvidia/os-interface.c
--- a/nvidia/os-interface.c	2025-12-08 14:45:03.000000000 +0100
+++ b/nvidia/os-interface.c	2026-01-27 11:35:52.318220971 +0100
@@ -1363,7 +1363,7 @@ NvU32 NV_API_CALL os_get_grid_csp_suppor
 
 void NV_API_CALL os_bug_check(NvU32 bugCode, const char *bugCodeStr)
 {
-    panic(bugCodeStr);
+    panic("%s",bugCodeStr);
 }
 
 NV_STATUS NV_API_CALL os_get_euid(NvU32 *pSecToken)
